micro-rdk/src/esp32/camera.rs

```diff
@@ -1,117 +1,254 @@
+        let xclk_freq_hz = cfg
+            .get_attribute::<i32>("xclk_freq_hz")
+            .unwrap_or(20_000_000);
+        let ledc_timer = cfg.get_attribute::<u32>("ledc_timer").unwrap_or(1);
+        let ledc_channel = cfg.get_attribute::<u32>("ledc_channel").unwrap_or(1);
+        let frame_size = cfg
+            .get_attribute::<u32>("frame_size")
+            .unwrap_or(FrameSize::QVGA as u32);
+        // Quality of JPEG output: 0-63 lower means higher quality
+        let jpeg_quality = cfg.get_attribute::<i32>("jpeg_quality").unwrap_or(32);
+        //  If pin_sccb_sda is -1, use the already configured I2C bus by number
+        let sccb_i2c_port = cfg.get_attribute::<i32>("sccb_i2c_port").unwrap_or(0);
+
+        let cam = Self {
             config: camera_config_t {
-                pin_pwdn: -1,
-                pin_reset: -1,
-                pin_xclk: 21,
-                __bindgen_anon_1: camera_config_t__bindgen_ty_1 { pin_sccb_sda: 26 },
-                __bindgen_anon_2: camera_config_t__bindgen_ty_2 { pin_sccb_scl: 27 },
-                pin_d7: 35,
-                pin_d6: 34,
-                pin_d5: 39,
-                pin_d4: 36,
-                pin_d3: 19,
-                pin_d2: 18,
-                pin_d1: 5,
-                pin_d0: 4,
-                pin_vsync: 25,
-                pin_href: 23,
-                pin_pclk: 22,
-                xclk_freq_hz: 20000000,
-                ledc_timer: 1,
-                ledc_channel: 1,
-                pixel_format: 4,
-                frame_size: 4,
-                jpeg_quality: 32,
+                pin_pwdn,
+                pin_reset,
+                pin_xclk,
+                __bindgen_anon_1: camera_config_t__bindgen_ty_1 { pin_sccb_sda },
+                __bindgen_anon_2: camera_config_t__bindgen_ty_2 { pin_sccb_scl },
+                pin_d7,
+                pin_d6,
+                pin_d5,
+                pin_d4,
+                pin_d3,
+                pin_d2,
+                pin_d1,
+                pin_d0,
+                pin_vsync,
+                pin_href,
+                pin_pclk,
+                xclk_freq_hz,
+                ledc_channel,
+                ledc_timer,
+                pixel_format: PixelFormat::JPEG as u32,
+                frame_size,
+                jpeg_quality,
+                // Number of frame buffers to be allocated.
+                // If more than one, then each frame will be acquired (double speed)
+                // when pixel_format == jpeg, fb_count > 1 goes to continuous mode, may need to adjust
+                // xclk_freq_hz down to 10_000_000.
                 fb_count: 1,
                 grab_mode: 0,
                 fb_location: 0,
-                sccb_i2c_port: 0,
+                sccb_i2c_port,
             },
-            last_grab: t.now(),
-        }
+        };
+
+        esp!(unsafe { esp_camera_init(&cam.config) })
```

Comment:
- This reads to me like there can only ever be one camera. If that is true, there probably should be some guardrails around trying to instantiate more than one. Or, have you tested this and esp32_camera does a good job handling that situation?
- The docs for `esp32_camera_init` state 'note call camera_probe before calling this function'. Is that being done somewhere? It also seems a little weird, because it looks like `esp32_camera_init` also calls that?

Reply:
I have not tested having multiple cameras.
There are a couple ways you could handle this, but I think deserves its own ticket.

1. since there's a limited number of pins and a single camera needs almost half of them, tracking which pins are currently allocated on the device and throwing an error when trying to assign them
2. setting a global to track if a camera has already been initialized, throwing an error if so.

On your second point, I see the comment in `esp32-camera` that refers to this, but as you said, the probe is called in the function. Within the examples, the probe function isn't called either.